### PR TITLE
chore: test apex checkpointing [DET-7886]

### PR DESCRIPTION
**Pull request by drh-determined-ai**
## Description

Test checkpointing and restoring with a simple linear model in an Apex or PyTorch AMP context.

## Test Plan

Run unit tests with GPU(s) and with `apex` installed.

## Commentary (optional)

Because these tests rely on GPU(s), they are **not currently being run by CircleCI**.

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 1]
```diff
@@ -229,6 +229,16 @@ def build_validation_data_loader(self) -> torch.utils.data.DataLoader:
             raise ValueError(f"unknown dataloader_type: {self.hparams['dataloader_type']}")
 
 
+class OneVarApexTrial(OneVarTrial):
```
Comment (rb-determined-ai) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:232]: Why is this distinct from `OneVarApexAMPTrial`?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 2]
```diff
@@ -657,6 +657,55 @@ def test_amp(self, trial_class) -> None:
...
+                expose_gpus=True,
+            )
+            controller.run()
+
+        for c in range(1, num_checkpoints):
+            prev_checkpoint_last_loss = training_metrics[c - 1][-1]["loss"]
+            this_checkpoint_first_loss = training_metrics[c][0]["loss"]
+            assert (
+                this_checkpoint_first_loss < prev_checkpoint_last_loss
+            ), f"loss did not decrease after checkpoint {c-1}"
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: I don't think this is a strong enough assertion.

It is true that the bug we're investigating is about "loss spikes after checkpoint restoring" or something.  But due to the linear model, stable inputs, and simple SGD loss function, I would be surprised to _ever_ see the loss increase, even if we totally messed up saving the scaler state.

Can you instead write the test so that it directly measures that the state of the scaler was properly saved/restored?  I can think of two ways to do it, either:
- direct inspection of scaler state, or
- inferring scaler state by expecting precise loss values, as with the last test you wrote

In either case, you'll need to make sure that the scaler finishes each run in a different state than its initial state, otherwise there's no way to tell if things were preserved or not.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py (hunk 1 above)
Comment (drh-determined-ai) [reply harness/tests/experiment/fixtures/pytorch_onevar_model.py:232]: I don't want the wacky data for this test as we had for the autocasting & scaling tests; I want to be able to expect the loss to decrease monotonically. The trial class names are not descript as they are right now.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 2 above)
Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: Anda had a similar concern in one of our calls, I think. What I can do is check if the scale value changed, and if it did check that there is a good, apparent reason (similar to what we did in the AMP tests).

An obvious problem we'd catch is that the scaler value resets to whatever initial value upon loading the state.

Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: Actually, scratch all that. That's just doing the same test as before. What I'm writing now and will push soon is this: the scale value just before saving a checkpoint should be the same as just after loading that checkpoint.

**Comment by drh-determined-ai** [comment]
Note that there is the existing test `TestPyTorchTrial.test_checkpointing_and_restoring`. An alternative approach to this task would be to parameterize the `OneVarTrial` subclass (similarly as is being done for `test_amp`).

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 3]
```diff
@@ -657,6 +657,61 @@ def test_amp(self, trial_class) -> None:
...
+        checkpoint_dir = str(tmp_path.joinpath("apex_checkpoint"))
+        latest_checkpoint = None
+        steps_completed = 0
+
+        def make_workloads(checkpoint_idx: int = 0) -> workload.Stream:
+            # Train as usual
+            trainer = utils.TrainAndValidate()
+            # In order to expect the loss to decrease after loading the first checkpoint,
+            #  make sure the scaler has had enough steps to stop reducing so that training
+            #  can actually proceed.
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:677]: how about just simplify the test so it doesn't make the "monotonic decreasing" assertion?  Because this is a unit test, it is reasonable that it only focuses on exactly one detail (preservation of scaler state, for instance).

The other test you wrote provides very good coverage of the loss behavior, but no coverage of scaler preservation around checkpoints.  This test could very reasonably provide the scaler preservation coverage without having to think about loss behavior.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 4]
```diff
@@ -657,6 +657,61 @@ def test_amp(self, trial_class) -> None:
...
+            yield from trainer.send(steps=10, validation_freq=10)
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:678]: nit: why even validate at all?  Validation should not affect the scaler at all.

Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:678]: I would agree, but `validation_freq` isn't an optional parameter, so I used the value that minimizes the time spent validating without needing clarification (like, I didn't put `11` or `inf`).

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 5]
```diff
@@ -657,6 +657,61 @@ def test_amp(self, trial_class) -> None:
         )
         controller.run()
 
+    @pytest.mark.skipif(not HAVE_APEX, reason="Apex not available")
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
+    @pytest.mark.gpu
+    def test_apex_checkpointing(self, tmp_path: pathlib.Path) -> None:
+        # There will be num_checkpoints-1 assertions made at the end of this test
+        num_checkpoints = 5
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:665]: question: what does num_checkpoints=5 buy us that num_checkpoints=2 wouldn't buy us?

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: that sounds good.  Can you also include an assertion to guarantee that the scaler state change over time?  I.e. that the test isn't passing on accident due to the scaler state always being the initial value.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:678]: huh, I didn't realize we couldn't not validate.  Makes sense then.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 2 above)
Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: I could, but that'll make the test resemble the AMP autoscaling test.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:677]: Because the scaler state changes during training, so reloading the same state over and over doesn't really prove anything

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:665]: Cool, so let's pick to the minimum num_checkpoints to achieve full coverage.  Otherwise we're just wasting time in tests.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:707]: I think avoiding false positives in tests is more important than the DRY principle.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 5 above)
Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py]: This code is gone.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py (hunk 2 above)
Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py]: We are now using the same assertion code as is being used in the AMP scaling tests.

File: harness/tests/experiment/fixtures/pytorch_onevar_model.py [hunk 6]
```diff
@@ -229,6 +229,25 @@ def build_validation_data_loader(self) -> torch.utils.data.DataLoader:
...
```
Comment (rb-determined-ai) [inline harness/tests/experiment/fixtures/pytorch_onevar_model.py:232]: OneVarApexTrial is no longer used, right?

Comment (drh-determined-ai) [reply harness/tests/experiment/fixtures/pytorch_onevar_model.py:232]: Ah, correct!

File: harness/tests/experiment/utils.py [hunk 7]
```diff
@@ -348,7 +348,10 @@ def make_workloads(steps: int) -> workload.Stream:
 
 
 def checkpointing_and_restoring_test(
```
Comment (rb-determined-ai) [inline harness/tests/experiment/utils.py:350]: I like the change to this function.  Very neat.

File: harness/tests/experiment/pytorch/test_pytorch_trial.py [hunk 8]
```diff
@@ -148,7 +148,35 @@ def make_workloads() -> workload.Stream:
...
+            pytorch_onevar_model.OneVarManualAMPTrial,
+        ],
+        ids=[
+            "default",
+            "apex",
+            "autocast",
+            "manual",
+        ],
+    )
+    def test_checkpointing_and_restoring(self, trial_class, tmp_path: pathlib.Path) -> None:
```
Comment (rb-determined-ai) [inline harness/tests/experiment/pytorch/test_pytorch_trial.py:167]: But this test is now very hard to read.  There are just too many if statements.  To me, that's a sign that the code paths are not actually similar enough to be combined.  Let's leave the original test alone and stick with a separate AMP test.

Comment (drh-determined-ai) [reply harness/tests/experiment/pytorch/test_pytorch_trial.py:167]: I agree 100%. I was able to tell you my same feelings!

**Review by rb-determined-ai — APPROVED** [review]
Great, I love this new test!

**Review by azhou-determined — APPROVED** [review]
looks great. this adds a needed test to our repository and also confirms that we don't need to fix that long-standing AMP checkpointing bug. you killed 2 JIRAs with one PR. 🥇
